src/ir/ir_traversing_visitor_with_result.h

```diff
@@ -0,0 +1,80 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#ifndef SRC_IR_IR_TRAVERSING_VISITOR_WITH_RESULT_H_
+#define SRC_IR_IR_TRAVERSING_VISITOR_WITH_RESULT_H_
+
+#include <variant>
+#include "src/ir/ir_visitor.h"
+#include "src/ir/module.h"
+
+namespace raksha::ir {
+
+// A visitor that also traverses the children of a node and allows performing
+// different actions before (PreVisit) and after (PostVisit) the children are
+// visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
+template <typename Derived, typename Result=std::monostate>
+class IRTraversingVisitorWithResult : public IRVisitor<Derived, Result> {
```

Comment:
Why do we have a separate class? Can't we just add an additional `Result` parameter to `IrTraversingVisitor`?

Reply:
Aferr suggested that I implement beside the IRTraversingVisitor (at least until incompatibilities can be resolved).

Reply:
After seeing this draft, I don't see any major incompatibilities. Let us go ahead and build it into IRTraversingVisitor itself?

Reply:
Ah, I hadn't sent this out before our chat today.

For completeness:
Sounds good to me :) I'll upload the split off PR as you suggested and see how easy the change is to merge in.
The main difference is relying on the default construction of the `Return` type to have a decent default for the `PreVisit` functions and defaulting `Return` to `std::monostate`.

R.E. Our chat, I'll upload the updated form soon.